### PR TITLE
Handle link with back/forward navigation

### PR DESCRIPTION
**Pull request by timneutkens**
## Bug

- [ ] Related issues linked using `fixes #number`
- [ ] Integration tests added
- [ ] Errors have helpful link attached, see `contributing.md`

## Feature

- [ ] Implements an existing feature request or RFC. Make sure the feature request has been accepted for implementation before opening a PR.
- [ ] Related issues linked using `fixes #number`
- [ ] Integration tests added
- [ ] Documentation added
- [ ] Telemetry added. In case of a feature if it's used or not.
- [ ] Errors have helpful link attached, see `contributing.md`

## Documentation / Examples

- [ ] Make sure the linting passes by running `pnpm lint`
- [ ] The examples guidelines are followed from [our contributing doc](https://github.com/vercel/next.js/blob/canary/contributing.md#adding-examples)

### CODE REVIEW

**Comment by ijjk** [comment]
## Failing test suites 
Commit: 870bf118d4ae18fb4d01d325291279340119f2d6

`yarn testheadless test/development/basic-basepath/hmr.test.ts` 
- basic HMR > Error Recovery > should recover from errors in getInitialProps in client

<details>
<summary>Expand output</summary>

  ● basic HMR › Error Recovery › should recover from errors in getInitialProps in client

    page.waitForSelector: Timeout 30000ms exceeded.
    =========================== logs ===========================
    waiting for selector "#error-in-gip-link"
    ============================================================

      304 |     return this.chain(() => {
      305 |       return page
    > 306 |         .waitForSelector(selector, { timeout, state: 'attached' })
          |          ^
      307 |         .then(async (el) => {
      308 |           // it seems selenium waits longer and tests rely on this behavior
      309 |           // so we wait for the load event fire before returning

      at lib/browsers/playwright.ts:306:10

</details>


Read more about building and testing Next.js in [contributing.md](https://github.com/vercel/next.js/blob/canary/contributing.md).

File: packages/next/client/components/app-router.client.tsx [hunk 1]
```diff
@@ -1,74 +1,61 @@
...
+  const [appRouter, previousUrlRef, current] = useRouter(initialUrl)
+
+  const onPopState = React.useCallback(
+    ({ state }: PopStateEvent) => {
+      if (!state) {
+        return
+      }
+      // @ts-ignore useTransition exists
+      // TODO: Ideally the back button should not use startTransition as it should apply the updates synchronously
+      React.startTransition(() => appRouter.replace(state.url))
```
Comment (SukkaW) [inline packages/next/client/components/app-router.client.tsx:43]: `startTransition` only exists in React 18. Will Next.js still support React 17 after the changes?

File: packages/next/client/components/userouter.ts [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+export default function useRouter(initialUrl: string): any {
+  const initialState = {
+    url: initialUrl,
+  }
+  const previousUrlRef = React.useRef(initialState)
+  const [current, setCurrent] = React.useState(initialState)
+  const change = React.useCallback(
+    (method: 'replaceState' | 'pushState', url: string) => {
+      // @ts-ignore startTransition exists
+      React.startTransition(() => {
```
Comment (SukkaW) [inline packages/next/client/components/userouter.ts:13]: https://github.com/vercel/next.js/pull/37332/files#r886331739

File: packages/next/client/link.tsx [hunk 3]
```diff
@@ -268,6 +277,8 @@ const Link = React.forwardRef<HTMLAnchorElement, LinkPropsReal>(
     }
 
     const p = prefetchProp !== false
+    // @ts-ignore useTransition exists
+    const [, /* isPending */ startTransition] = React.useTransition()
```
Comment (SukkaW) [inline packages/next/client/link.tsx:281]: https://github.com/vercel/next.js/pull/37332/files#r886331739

Comment (timneutkens) [reply packages/next/client/link.tsx]: Added an extra check to this one to ensure it's not called with React 17, good catch 👍

File: packages/next/client/components/userouter.ts (hunk 2 above)
Comment (timneutkens) [reply packages/next/client/components/userouter.ts:13]: https://github.com/vercel/next.js/pull/37332#discussion_r886591244

**Review by shuding — APPROVED** [review]

**Pull request merged**